pageserver/src/walredo.rs

```diff
@@ -587,419 +583,1506 @@ impl<C: CommandExt> CloseFileDescriptors for C {
+                None => {
+                    // the Handle or the request queue sender have been dropped; return Ok(()) to keep
+                    // processing any of already pipelined requests.
+                    //
+                    // the drop however does never seem to happen, but most likely the task is
+                    // descheduled because of a shutdown_watcher.
+                    return Ok(());
+                }
             }
+        };
 
-            // If we have some messages in stderr, forward them to the log.
-            let err_revents = pollfds[1].revents().unwrap();
-            if err_revents & (PollFlags::POLLERR | PollFlags::POLLIN) != PollFlags::empty() {
-                let mut errbuf: [u8; 16384] = [0; 16384];
-                let n = self.stderr.read(&mut errbuf)?;
+        let timeout_at = tokio::time::Instant::now() + request.timeout;
 
-                // The message might not be split correctly into lines here. But this is
-                // good enough, the important thing is to get the message to the log.
-                if n > 0 {
-                    error!(
-                        "wal-redo-postgres: {}",
-                        String::from_utf8_lossy(&errbuf[0..n])
-                    );
+        let records = request.records_range.sub_slice(&request.records);
+
+        if have_vectored_stdin {
+            build_vectored_messages(
+                request.target,
+                request.base_img,
+                records,
+                &mut scratch,
+                &mut buffers,
+            );
+        } else {
+            build_messages(
+                request.target,
+                request.base_img,
+                records,
+                &mut scratch,
+                &mut buffers,
+            );
+        }
+
+        let write_res = async {
+            while buffers.has_remaining() {
+                poll_fn(|cx| poll_write_buf(Pin::new(&mut stdin), cx, &mut buffers))
+                    .await
+                    .map_err(anyhow::Error::new)?;
+            }
+            // in general flush is not needed, does nothing on pipes, but since we now accept
+            // AsyncWrite
+            stdin.flush().await.map_err(anyhow::Error::new)
+        };
 
-                    // To make sure we capture all log from the process if it fails, keep
-                    // reading from the stderr, before checking the stdout.
-                    continue;
+        let write_res = tokio::time::timeout_at(timeout_at, write_res)
+            .map_err(|_| anyhow::anyhow!("write timeout"));
+
+        // wait the write to complete before sending the completion over, because we cannot fail
+        // the request after it has been moved. this could be worked around by making the oneshot
+        // into a normal mpsc queue of size 2 and making the read side race against the channel
+        // closing and timeouted read.
+        match write_res.await.and_then(|x| x) {
```

Comment:
What is the deal with this `.and_then(|x| x)?

I think it wouldn't hurt to have a preceding line like

`let write_res = write_res.await.and_then(|x| x); // explain why`

Reply:
I use `.and_then(|x| x)` for [`Result::flatten`](https://doc.rust-lang.org/stable/std/result/enum.Result.html#method.flatten) (still unstable, unfortunately).

So the timeout returns `Result<Inner, tokio::time::Elapsed>` which I `map_err` into `anyhow::Result<Inner>`, but the `Inner` is `anyhow::Result<()>`, so flattening gives me one `anyhow::Result<()>`.